server/coordinator/scheduler/nodepicker/hash/consistent_uniform.go

```diff
@@ -126,24 +140,37 @@ func BuildConsistentUniformHash(numPartitions int, members []Member, config Conf
 
 	numReplicatedNodes := len(members) * config.ReplicationFactor
 	avgLoad := float64(numPartitions) / float64(len(members))
+	minLoad := int(math.Floor(avgLoad))
+	maxLoad := int(math.Ceil(avgLoad))
+
+	memPartitions := make(map[string]map[int]struct{}, len(members))
+	for _, mem := range members {
+		memPartitions[mem.String()] = make(map[int]struct{}, maxLoad)
+	}
+
+	// Sort the affinity rule to ensure consistency.
+	sort.Slice(config.PartitionAffinities, func(i, j int) bool {
+		return config.PartitionAffinities[i].PartitionID < config.PartitionAffinities[j].PartitionID
+	})
 	c := &ConsistentUniformHash{
 		config:        config,
-		minLoad:       math.Floor(avgLoad),
-		maxLoad:       math.Ceil(avgLoad),
+		minLoad:       minLoad,
+		maxLoad:       maxLoad,
 		numPartitions: uint32(numPartitions),
 		sortedRing:    make([]virtualNode, 0, numReplicatedNodes),
+		memPartitions: memPartitions,
 		members:       make(map[string]Member, len(members)),
-		memLoads:      make(map[string]float64, len(members)),
 		partitionDist: make(map[int]int, numPartitions),
 		nodeToMems:    make(map[virtualNode]Member, numReplicatedNodes),
 	}
 
 	c.initializeVirtualNodes(members)
 	c.distributePartitions()
+	c.ensureAffinity()
```

Comment:
What will happen If the affinity rules cannot be satisfied?
For example:
There are now 10 shards and 2 nodes, but the NumAllowedOtherPartitions of 3 shards are set to 0.
Is it necessary to do a pre-verification of affinity rules?

Reply:
Actually, some affinity rule may not be met if so, but it will stop.